### PR TITLE
Initial GOV.UK PAAS guide

### PR DESCRIPTION
**Pull request by saliceti**
Adding suitability section for paas

### CODE REVIEW

File: guides/default-technology-stack.md [hunk 1]
```diff
@@ -47,15 +47,11 @@ Community support for Azure use in general can also be gained from the community
 
 ### GOV.UK PaaS
 
-We are currently exploring usage of [GOV.UK PaaS](https://www.cloud.service.gov.uk/) as a secondary offering, particularly for [12 factor apps](https://12factor.net/).
+We are currently exploring usage of [GOV.UK Platform as a Service](https://www.cloud.service.gov.uk/) as a secondary offering for applications requiring less customisation than infrastructure as a service like Azure.
```
Comment (himal-mandalia) [inline guides/default-technology-stack.md:50]: "a secondary offering for applications requiring less customisation than provided by a full infrastructure-as-a-service platform such as Azure"

Think that's clearer

File: guides/default-technology-stack.md [hunk 2]
```diff
@@ -47,15 +47,11 @@ Community support for Azure use in general can also be gained from the community
...
 
-* [Features](https://www.cloud.service.gov.uk/features)
-* [Documentation](https://docs.cloud.service.gov.uk/)
-* [Pricing](https://www.cloud.service.gov.uk/pricing)
+Several production applications are already running on GOV.UK PaaS but it still needs to be approved by the DfE architecture and security teams on a case-by-case basis. This is currently being progressed.
```
Comment (himal-mandalia) [inline guides/default-technology-stack.md:52]: Can add "Please speak with your architecture partner or contact architecture.services@education.gov.uk"

File: guides/govuk-paas.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+---
+category: Development Guides
+expires: 2020-05-01
+---
+
+# GOV.UK Platform as a Service
+
+Gov.uk Platform as a Service (PaaS) is a platform developed by [Government Digital Service (GDS)](https://www.gov.uk/government/organisations/government-digital-service) and available for Government departments to deploy applications and backend services.
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:8]: GOV.UK instead of Gov.uk - to stay consistent with style guide

Comment (himal-mandalia) [reply guides/govuk-paas.md:8]: "Government departments" - no need to capitalise the G in government here

File: guides/govuk-paas.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+It simplifies deploying to the cloud as a lot of the requirement for production applications is already implemented: internet access, routing, container scheduling, scalability, observability, redundancy, logging, monitoring, user access management, billing...
+
+It works well for a number of use cases. When more customisation is required, it is recommended to use [Azure CIP](/guides/default-technology-stack/#azure).
+
+### 12-factor apps
+PaaS is mainly aimed at [12-factor applications](https://12factor.net/), which is a set of good practices for web applications to ensure smooth operations, scalability and consistency across environment.
+
+### Security
+The platform is assured for use at OFFICIAL. As the data owner for a service [you must ensure that appropriate controls are in place when handling Official data](https://www.gov.uk/guidance/official-sensitive-data-and-it).
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:20]: Where it say OFFICIAL you can link to this: https://www.cloud.service.gov.uk/security/

File: guides/govuk-paas.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+### Containers
+It runs applications in stateless containers, horizontally scalable. It is not adapted to large stateful entreprise applications.
+
+### Internet facing
+The platform runs in the cloud and is exclusively accessed via Internet for inbound and outbound connections. It can't be connected via dedicated network links or use IP restrictions.
+
+### Backing services
+A limited number of backing services are provided: postgres, mysql, redis, elasticsearch, AWS S3, InfluxDB. They run in the platform and cannot be accessed from outside.
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:29]: Can link to this page here too: https://www.cloud.service.gov.uk/application-development/#backing-services

**Review by himal-mandalia — COMMENTED** [review]
Some minor amends and links to add

**Comment by misaka** [comment]
Just on the TA call and the comments here are that we need to make it clearer how and when it would make sense to use GOVUK PAAS over other offerings such as CIP. Some of the comments made there are:

From @ddippolito 

> These are two use cases:
>
> GOV.UK PaaS proves to be a valid candidate for services that don’t have complex needs in terms of infrastructure customisation. 
>
> It’s also a sensible choice for teams who want to experiment during alpha prototyping exercises or just moving into private beta as it allows them to be more flexible and independent without having to rely on infrastructure ops engineers to spin up new applications. 
>
> Taken from that doc I wrote a while ago

From @himal-mandalia

> I covered this in a slide deck last year too: 
>
> https://docs.google.com/presentation/d/10E_Ej4MsU5Uhyj6dppdd4IZklGz9ymmF8brWTZP0x_I/edit#slide=id.p


From @robhettrick:

> Might be useful framework for decision making here, about to Gov.Uk PaaS or not https://docs.google.com/spreadsheets/d/1BiknOwpzpTXX_ke_eAeWT8v5ZqzOt8BpZLkZdfEhNQ8/edit?usp=drivesdk

**Comment by himal-mandalia** [comment]
Thanks for capturing @misaka 

Also to add to the point from @ddippolito - not only does it allow for ease of hosting for prototyping in alpha, it works out much cheaper than Heroku which is what most teams use for these exercises.

File: guides/default-technology-stack.md (hunk 1 above)
Comment (misaka) [inline guides/default-technology-stack.md:50]:
```suggestion
We are currently exploring usage of [GOV.UK Platform as a Service](https://www.cloud.service.gov.uk/) as a secondary offering for applications requiring less customisation than provided by a full infrastructure-as-a-service platform such as Azure. It is suitable for less complex web services that follow typical GOV.UK patterns.
```

File: guides/govuk-paas.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
+---
+category: Development Guides
+expires: 2020-05-01
+---
+
+# GOV.UK Platform as a Service
+
+GOV.UK Platform as a Service (PaaS) is a platform developed by [Government Digital Service (GDS)](https://www.gov.uk/government/organisations/government-digital-service) and available for government departments to deploy applications and backend services. It is based on the widely used open source platform [Cloud Foundry](https://www.cloudfoundry.org/).
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:8]: Expand here:

> It is based on the widely used open source platform Cloud Foundry and built on Amazon Web Services.

File: guides/govuk-paas.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+Several production applications are already running on GOV.UK PaaS but it still needs to be approved by the DfE architecture and security teams on a case-by-case basis. This is currently being progressed. In the meantime, please speak with your architecture partner or contact architecture.services@education.gov.uk.
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:10]: Reword to make language clear:

> Several production applications are already running on GOV.UK PaaS but production usage still needs to be approved by the DfE architecture and security teams on a case-by-case basis. This is currently being progressed. In the meantime, please speak with your architecture partner or contact architecture.services@education.gov.uk.

File: guides/govuk-paas.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+## Suitability
+
+[Azure CIP](/guides/default-technology-stack/#azure) is the default platform for new projects in DfE. When does it make sense to use GOV.UK PaaS?
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:14]: > is the strategic and default platform for new hosting in DfE

File: guides/govuk-paas.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+### Standard web applications
+
+GOV.UK PaaS runs applications in stateless Linux containers, horizontally scalable. It is not adapted to large stateful entreprise applications for instance.
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:18]: > GOV.UK PaaS runs applications in stateless Linux containers and is horizontally scalable. It is not adapted to large stateful enterprise applications.

File: guides/govuk-paas.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+It is mainly aimed at [12-factor applications](https://docs.cloud.service.gov.uk/architecture.html#12-factor-application-principles), which is a set of good practices for web applications to ensure smooth operations, scalability and consistency across environments. The service manual [recommends](https://www.gov.uk/service-manual/technology/deploying-software-regularly#managing-variable-configuration) this architecture.
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:20]: > It is mainly aimed at [12-factor applications](https://docs.cloud.service.gov.uk/architecture.html#12-factor-application-principles), which is a set of principles and good practice for web applications which ensure smooth operation, scalability and consistency across environments. The service manual [recommends](https://www.gov.uk/service-manual/technology/deploying-software-regularly#managing-variable-configuration) this architecture.

File: guides/govuk-paas.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+Native Windows applications will not run. However, .NET Core is supported as it runs on Linux.
+
+### Rapid development
+
+The target platform may be Azure, but it may take time to set up and find the right engineers to manage it. PaaS' ability to spin up and down environments by anyone enables prototyping and fast iterations in Discovery or Alpha stages. Some teams use [Heroku](https://heroku.com/) to deploy a prototype quickly. PaaS provides similar features but cheaper and hosted by Government.
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:26]: > The strategic Azure platform may take time and resource to set up, requiring specialised engineers to set up and manage. PaaS provides the ability to spin environments up/down without infrastructure expertise, enabling prototyping and fast iteration, which is particularly useful in Discovery and Alpha. Some teams use [Heroku](https://heroku.com/) to deploy a prototype quickly. PaaS provides similar features but government hosted and cheaper.

File: guides/govuk-paas.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+Also, it allows short lived environments created for the unique purpose of validating a new feature for example, similarly to [Heroku review apps](https://devcenter.heroku.com/articles/github-integration-review-apps). This removes the contention usually observed when sharing environments.
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:28]: > Also, it allows short lived environments created for the unique purpose of validating a new feature, similar to [Heroku review apps](https://devcenter.heroku.com/articles/github-integration-review-apps). This allows for ephemeral/temporary environments to be created and disposed "on the fly" and removes the contention usually observed when sharing environments.

File: guides/govuk-paas.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+### Low cost
+
+PaaS is particularly adapted to services with budget constraints as total cost is lower than full IaaS. Cost of hosting is usually lower than Azure but most of the savings will come from the lower dependency on dedicated infrastructure engineers.
+
+Specialist engineers may still be requested on an ad hoc basis for guidance and to make sure the application is production ready.
+
+### Simple infrastructure
+
+Setting up a web application in the cloud involves a significant number of non-funcional requirements like providing internet access, container scheduling, load balancing, scalability, monitoring, user access management, billing...
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:38]: > Setting up a web application in the cloud involves a significant number of non-funcional requirements such as providing internet access, container scheduling, load balancing, scalability, monitoring, user access management, billing etc.

File: guides/govuk-paas.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+A platform-as-a-service [includes them already](https://docs.cloudfoundry.org/concepts/overview.html) so the developer doesn't have to implement them.
+
+As a result, setting up an application does not require full time infrastructure specialists or ops engineers. However, the set-up is limited to what is available in the platform: see [application development](https://www.cloud.service.gov.uk/application-development/) and [constraints](https://www.cloud.service.gov.uk/constraints/).
+
+A limited number of [backing services](https://www.cloud.service.gov.uk/application-development/#backing-services) are provided: postgres, mysql, redis, elasticsearch, AWS S3, InfluxDB. They run in the platform and cannot be accessed from outside.
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:43]: > A limited number of [backing services](https://www.cloud.service.gov.uk/application-development/#backing-services) are provided: PostgreSQL, MySQL, Redis, Elasticsearch, AWS S3, and InfluxDB. They run in the platform and cannot be accessed externally but can be easily linked to your application in a PaaS container.

File: guides/govuk-paas.md [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+Application logs and performance metrics are [available in the platform](https://docs.cloud.service.gov.uk/monitoring_apps.html#monitoring-apps) in standard format and can be accessed manually, via the web portal or via external tools. The platform itself doesn't provide log storage or custom visualisations.
+
+### Security: Up to OFFICIAL data
+
+The platform is [assured for use at OFFICIAL](https://www.cloud.service.gov.uk/security/#our-safety-measures). As the data owner for a service [you must ensure that appropriate controls are in place when handling Official data](https://www.gov.uk/guidance/official-sensitive-data-and-it).
+
+The platform is multi tenant: the resources are shared between different users and each one has dedicated access to the resources. Some very specific applications might require total isolation which the platform does not provide.
```
Comment (himal-mandalia) [inline guides/govuk-paas.md:51]: > The platform is multi tenant: the resources are shared between different users and each one has dedicated access to the resources, with clear access control policies. Users are other government departments. Some very specific applications might require total isolation which the platform does not provide.

File: guides/govuk-paas.md [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+### Connectivity: Internet only
+
+The platform runs in a public cloud: inbound and outbound connections are exclusively via public Internet. Traditional network based security via dedicated network links, VPN, IP restrictions cannot be used.
+
+Instead, transport security is achieved by using HTTPS and authentication via login/password, single-sign-on, OAuth...
```
Comment (himal-mandalia) [inline guides/govuk-paas.md]: > Instead, transport security is achieved by using HTTPS/TLS and authentication via username/password or a single-sign-on integration such as with OAuth or SAML, or several types of API authentication. This is handled at the application level. This is a "cloud first" approach as specified in the Technology Code of Practice.

**Review by himal-mandalia — COMMENTED** [review]
Just some minor language/wording amendments.

**Review by himal-mandalia — APPROVED** [review]
Looks good!

**Pull request merged**